Tests/E2E/MacTest.cs

```diff
@@ -204,5 +204,139 @@ private async Task Test_OTAA_Unconfirmed_Send_And_Receive_C2D_Mac_CommandsImplAs
+        {
+            const int messagesToSend = 10;
+            const int warmUpMessageCount = 2;
+
+            var device = TestFixtureCi.Device32_ABP;
+            LogTestStart(device);
+
+            // Setup LoRa device properties
+            await ArduinoDevice.setDeviceModeAsync(LoRaArduinoSerial._device_mode_t.LWABP);
+            await ArduinoDevice.setIdAsync(device.DevAddr, device.DeviceID, device.AppEUI);
+            await ArduinoDevice.setKeyAsync(device.NwkSKey, device.AppSKey, device.AppKey);
+
+            // Setup protocol properties
+            // Start with DR5
+            await ArduinoDevice.SetupLora(TestFixtureCi.Configuration.LoraRegion, LoRaArduinoSerial._data_rate_t.DR5, 4, true);
+            await TestFixture.CleanupC2DDeviceQueueAsync(device.DeviceID);
+
+            // Sends 2x unconfirmed messages
+            for (var i = 1; i <= warmUpMessageCount; ++i)
+            {
+                var msg = PayloadGenerator.Next().ToString(CultureInfo.InvariantCulture);
+                Log($"{device.DeviceID}: Sending unconfirmed '{msg}' {i}/{messagesToSend}");
+                await ArduinoDevice.transferPacketAsync(msg, 10);
+
+                await Task.Delay(Constants.DELAY_BETWEEN_MESSAGES);
+                await AssertUtils.ContainsWithRetriesAsync("+MSG: Done", ArduinoDevice.SerialLogs);
+                TestFixture.ClearLogs();
+            }
+
+            var c2dMessage = new LoRaCloudToDeviceMessage()
+            {
+                Fport = 0,
+                Payload = string.Empty,
+                MacCommands = { new LinkADRRequest(3, 4, 25, 0, 1) } // Update data rate to DR3
+            };
+
+            await TestFixtureCi.SendCloudToDeviceMessageAsync(device.DeviceID, c2dMessage);
+            Log($"C2D Message sent to device, need to check if it receives");
+
+            var foundC2DMessageCount = 0;
+            var foundLinkADRCmdCount = 0;
+            var foundChangedDataRateCount = 0;
+
+            var c2dLogMessage = $"{device.DeviceID}: cloud to device MAC command LinkADRCmd received Type: LinkADRCmd Answer, datarate: 3";
+            var linkADRCmdMessage = "LinkADRCmd mac command detected in upstream payload";
+            var changedDataRateMessage = $"\"datr\":\"SF9BW125\""; // DR3
+
+            // Sends 8x confirmed messages, stopping if C2D message is found and data rate is updated
+            for (var i = warmUpMessageCount + 1; i <= messagesToSend; ++i)
+            {
+                var msg = PayloadGenerator.Next().ToString(CultureInfo.InvariantCulture);
+                Log($"{device.DeviceID}: Sending unconfirmed '{msg}' {i}/{messagesToSend}");
+                await ArduinoDevice.transferPacketAsync(msg, 10);
+
+                await Task.Delay(Constants.DELAY_BETWEEN_MESSAGES);
+
+                // After transferPacket: Expectation from serial
+                // +MSG: Done
+                await AssertUtils.ContainsWithRetriesAsync("+MSG: Done", ArduinoDevice.SerialLogs);
```

Comment:
Does the Arduino print something to the serial for the acknowledgement for the DR change? Or would it be too brittle to assert such a serial log message?

Reply:
Not that I have seen. The C2D message seems to be acknowledged just like any other message.